### ISSUE TITLE
942300 matching "for ? (" when it's expected to find OR?

### ISSUE THREAD

**Issue by Rmh-Tech** [coreruleset/coreruleset#2007]
### Description

I have a small number of false positive on CRS 3.1.0 rule ID 942300, using Application Gateway in Azure so logs are not available in full/normal standard but this does not seem like intentional blocking.

It seems the rule matches incorrectly with the following examples:
"pay me for holiday(tomorrow"
"booked for 2021(including 2020"
"proof of their earnings and savings(if any)."

Maybe it's just the char injection issue ?(? but I didn't expect the word FOR to match because it ends with OR

--

### Audit Logs / Triggered Rule Numbers

942300

### Your Environment

* CRS version (e.g., v3.2.0): 3.1.0
* Paranoia level setting: n/a Azure Application Gateway
* ModSecurity version (e.g., 2.9.3): ?
* Web Server and version (e.g., apache 2.4.41): IIS?
* Operating System and version: IIS? WS?

### Confirmation

[Y] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by franbuehler** [comment]
Issue chat meeting Feb 15: @csanders-git volunteers to take this one.

**Comment by dune73** [comment]
@csanders-git : Did you have time to look into this? If there is a status, please share.

**Comment by franbuehler** [comment]
Since @csanders-git has assigned this issue for a longer time, I took a look here.

The false positive appears because of the `OR` in FOR.
The FPs could easily be solved with a word boundary.

Current regex snippet: `or\s+\s*?\w+\(`
Proposal: `\bor\s+\s*?\w+\(`

Should I propose a PR here?

**Comment by fzipi360** [comment]
What about the last sentence? "proof of their earnings and savings(if any)."

**Comment by fzipi360** [comment]
Does this `\s*?` make sense also in that regexp? I think `\s*` is the same as `\s*?`, in practical terms.

**Comment by franbuehler** [comment]
Thank you @fzipi :-)

> What about the last sentence? "proof of their earnings and savings(if any)."
> I think it is the same problem, but for and\s+\s*?\w+\(, right?

Yes, absolutely. I overlooked this one... Yes, my solution does not work for this example. :(

> Does this \s*? make sense also in that regexp? I think \s* is the same as \s*?, in practical terms.

Yes, almost. `\s*` is greedy and `\s*?` is non greedy or lazy. 
I don't understand why a `\s+` is followed by a `\s*?`
It's the same as `\s+?`, isn't it??

To summarize it, my solution does not work for both mentioned false positives...

**Comment by fzipi** [comment]
Ups, sorry, my other user commented before :facepalm: 
It should be exactly the same as `\s+?`, agreed.

But your solution works, but we add the same for and: `\band\s+\s*?\w+\(`

**Comment by franbuehler** [comment]
I've just seen that I already started with the PR here. Is it ok when I finish it and you take another one, @fzipi ??

**Comment by dune73** [comment]
That must be a pleasant surprise. For me it's usually the other way around. I think I have started with a PR, but when I return to it, it was only a surreal fantasy.

**Comment by franbuehler** [comment]
I didn't finish it because I was not able to test it with FTW locally.
I think it's time to test https://github.com/fzipi/go-ftw??

**Comment by fzipi** [comment]
Sure. JFYI, I was adding more verbs in the file. This is what I addded:
```
diff --git a/util/regexp-assemble/regexp-942300.data b/util/regexp-assemble/regexp-942300.data
index d35b538..f83b69a 100644
--- a/util/regexp-assemble/regexp-942300.data
+++ b/util/regexp-assemble/regexp-942300.data
@@ -3,16 +3,19 @@
 [\"'`]\s*?--
 [\"'`]\s*?{
 \/\*!\s?\d+
-char\s*?\(\s*?\d
-chr\s*?\(\s*?\d
-and\s+\s*?\w+\(
-nand\s+\s*?\w+\(
-or\s+\s*?\w+\(
-xor\s+\s*?\w+\(
-xxor\s+\s*?\w+\(
-div\s+\s*?\w+\(
-like\s+\s*?\w+\(
-between\s+\s*?\w+\(
-not\s+\s*?\w+\(
+\bbinary\s*?\(\s*?\d
+\bchar\s*?\(\s*?\d
+\bchr\s*?\(\s*?\d
+\band\s+\s*?\w+\(
+\bnand\s+\s*?\w+\(
+\bor\s+\s*?\w+\(
+\bxor\s+\s*?\w+\(
+\bxxor\s+\s*?\w+\(
+\bdiv\s+\s*?\w+\(
+\blike\s+\s*?\w+\(
+\bbetween\s+\s*?\w+\(
+\bnot\s+\s*?\w+\(
+\bregexp\s+\s*?\w+\(
+\brlike\s+\s*?\w+\(
 \|\|\s+\s*?\w+\(
 \&\&\s+\s*?\w+\(
```
I've realised that they looked like a small subset,  after taking a look at https://dev.mysql.com/doc/refman/5.7/en/non-typed-operators.html

**Comment by franbuehler** [comment]
Oh, wow, you started too! Sorry.
So, do you want to finish it and I'm trying to get your go-ftw up and running??

**Comment by fzipi** [comment]
Go ahead an try go-ftw!

I was on the writing tests stage ;)

**Comment by franbuehler** [comment]
I already had them. But I gave up with ftw :-(
